tubular/dates.py

```diff
@@ -14,132 +13,196 @@
 
-        self._generate_is_datetime64_dict(X)
+    def check_columns_are_date_or_datetime(
+        self,
+        X: pd.DataFrame,
+        datetime_only: bool,
+    ) -> None:
+        """Raise a type error if a column to be operated on is not a datetime.datetime or datetime.date object
 
-        for col in self.columns:
-            if (
-                not self._is_datetime64_dict[col]
-                and pd.api.types.infer_dtype(X[col]) != "date"
-            ):
-                msg = f"{self.classname()}: {col} should be datetime64 or date type but got {X[col].dtype}"
-                raise TypeError(msg)
+        Parameters
+        ----------
+
+        X: pd.DataFrame
+            Data to validate
+
+        datetime_only: bool
+            Indicates whether ONLY datetime types are accepted
 
-    def cast_columns_to_datetime(self, X: pd.DataFrame) -> pd.DataFrame:
-        """
-        Check whether columns are datetime64 or date type. If not, an error will be raised. If one is datetime.date
-        it will be cast to datetime64.
         """
 
-        self.check_columns_are_date_or_datetime(X)
+        type_dict = {}
+        datetime_type = "datetime64"
+        date_type = "date"
+        allowed_types = [datetime_type]
+        if not datetime_only:
+            allowed_types = [*allowed_types, date_type]
 
-        temp = X[self.columns]
+        for col in self.columns:
+            if pd.api.types.is_datetime64_any_dtype(X[col]):
+                type_dict[col] = datetime_type
 
-        for column in self.columns:
-            if not self._is_datetime64_dict[column]:
-                temp[column] = pd.to_datetime(X[column])
+            elif (not datetime_only) and (
+                pd.api.types.infer_dtype(X[col]) == date_type
+            ):
+                type_dict[col] = date_type
 
-                warnings.warn(
-                    f"""
-                    {self.classname()}: temporarily cast {column} from datetime64 to date before transforming in order to apply the datetime method.
+            else:
+                msg = f"{self.classname()}: {col} type should be in {allowed_types} but got {X[col].dtype}"
+                raise TypeError(msg)
 
-                    This will artificially increase the precision of each data point in the column. Original column not changed.
-                    """,
-                    stacklevel=2,
-                )
+        present_types = set(type_dict.values())
 
-        return temp
+        only_datetime_present = present_types == {datetime_type}
+        only_date_present = present_types == {date_type}
+        date_allowed = ~datetime_only
 
-    def _cast_non_matching_columns(
+        if not only_datetime_present and not (date_allowed and only_date_present):
+            msg = f"{self.classname()}: Columns fed to datetime transformers should be {allowed_types} and have consistent types, but found {present_types}. Please use ToDatetimeTransformer to standardise."
```

Comment:
P - good call on dropping the recasting and pointing to ToDateTimeTransformer.  Much simpler.